code/go/0chain.net/smartcontract/stakepool/stakepool.go

```diff
@@ -302,6 +322,38 @@ func (sp *StakePool) Empty(sscID, poolID, clientID string, balances cstate.State
 	return nil
 }
 
+// SlashFraction
+// slash stake pools funds, if a provider is killed
+func (sp *StakePool) SlashFraction(
+	killSlashFraction float64,
+	providerId string,
+	providerType spenum.Provider,
+	balances cstate.StateContextI,
+) error {
+	if killSlashFraction == 0.0 {
+		return nil
+	}
+	if killSlashFraction < 0 || killSlashFraction > 1 {
+		return fmt.Errorf("kill slash %v should be in the interval [0,1]", killSlashFraction)
+	}
+	reduction := 1 - killSlashFraction
+	if reduction < 0 {
+		reduction = 0
+	}
+	if reduction > 1 {
+		reduction = 1
+	}
+	for _, dp := range sp.Pools {
```

Comment:
We'd better process pools in order to avoid potential state mismatch. We can use `sp.OrderedPoolIds` to get ids first, get pool by id and process it.

Reply:
`currency.MultFloat64()` could return two kinds of error, the `ErrNegativeValue` and `ErrUint64OverflowsFloat64`. So if any chance miner1 got `ErrNegativeValue` first and `miner2 got the other one first, then we will see different MPT state. We MUST execute code in order and avoid random MAP iterating. Also see issue https://github.com/0chain/0chain/issues/2085#issuecomment-1421697224. 

We will also exame all the SCs to see if there're other places that do this random MAP accessing, so before fixing exist ones, let's not add new one.

Reply:
Ok makes sense, but these problems are not related to this PR, but should be handled in a separate issue.